Test Runner

gabi-250 opened the pull request:
This branch contains a very basic testing mechanism. `run_tests.py` is a python script that compiles the test files under `src/tests`, runs the resulting binaries, and compares their output with the expected output. This approach is obviously very limited, but it's helping me prevent regressions from happening.

I've also refactored `guard.c` (now it only contains the `__guard_failure` callback and a global variable which needs to be visible in `jump.s`).

I have just realized `libunwind`'s `unw_set_reg` might not actually be modifying any register values (I think it's just changing them in the internal structures used by `libunwind`). I need to investigate this further.

vext01 commented on src/stackmap_checker/call_stack_state.c:
```diff
@@ -0,0 +1,218 @@
+    free(state);
+}
+
+size_t get_locations(stack_map_t *sm, call_stack_state_t *state,
```
What does this function do?

vext01 commented on src/stackmap_checker/call_stack_state.c:
```diff
@@ -0,0 +1,218 @@
+    return state;
+}
+
+void collect_map_records(call_stack_state_t *state, stack_map_t *sm)
```
Similarly, let's comment up all medium to large functions so we know what they do 6 months down the line.

vext01 commented on src/stackmap_checker/call_stack_state.c:
```diff
@@ -0,0 +1,218 @@
+    free(locations);
+}
+
+void append_record(call_stack_state_t *state, stack_map_record_t first_rec)
```
Append (stackmap?) record to what?

vext01 commented on src/stackmap_checker/call_stack_state.h:
```diff
@@ -0,0 +1,63 @@
+
+void free_call_stack_state(call_stack_state_t *state);
+
+/*
```
Ah I see, you have documented what the functions do in the headers. I can roll with that.

vext01 commented on src/stackmap_checker/utils.c:
```diff
@@ -3,13 +3,26 @@
 {
-    int fd = open(TRACE_BIN, O_RDONLY);
+    char *buff = malloc(MAX_BUF_SIZE);
+    int buff_size = readlink("/proc/self/exe", buff, MAX_BUF_SIZE - 1);
```
I'm not too fussed about this, as this code is highly experimental, but not all systems have `/proc`. In a production system you'd have to look at argv[0].

gabi-250 replied:
I am aware of that, but there is no `argv` in my program, since there is no `main`. This is the only solution I could come up with.

vext01 commented:
This looks good, but, as we discussed the other day, these PRs are so large it's hard to do a proper review.

gabi-250 commented:
@vext01 Thank you for your review!

I've documented all the non-obvious functions in the header files. Do I need to address anything else before I squash?

gabi-250 commented:
@vext01 My bad, I forgot to add the python test runner.

vext01 commented on src/run_tests.py:
```diff
@@ -0,0 +1,52 @@
+    subprocess.run('cd {:} && make clean && make'.format(args.test_dir),
+                   shell=True, stdout=subprocess.PIPE)
+    c_files = extract_files_with_extension(args.test_dir, '.c')
+    bin_names = [name[:name.rfind('.')] for name in c_files]
```
You could simplify this with `os.path.splitext`:
https://docs.python.org/2/library/os.path.html#os.path.splitext

gabi-250 replied:
Good point!

gabi-250 commented:
@vext01 Thank you for your review! 

I've addressed your comment about the test runner. I am going to improve the tests and the runner in a future PR, as I am currently working on fixing an annoying bug.

vext01 commented:
Is this ready to squash?

gabi-250 commented:
@vext01 Squashed
